### PR TITLE
Editor switch fix

### PR DESCRIPTION
**Pull request by joyously**
## Description
This is an alternate approach to fixing the problem with the editor losing line breaks when switching modes.
It removes the change (again) from https://core.trac.wordpress.org/changeset/43336 that is specific to WP blocks.
It checks for existence of the selection marker before removing the elements.
It does not remove the marker parent (which is what deletes content).
It enhances `removep` and `autop` to preserve comments, pre, and script tags.
It includes a fix for having the cursor in a comment, from https://core.trac.wordpress.org/ticket/45947

## Motivation and context
The upgrade of the editor to 4.9.11 is desirable, but the change for WP blocks is not. This fixes the problems better than other attempts (which I'm not sure the state of).

## How has this been tested?
I tested locally, using the following data in a post, switching between Text and Visual modes with the cursor in various places. It's important to switch multiple times and to start the editor in both modes because different code paths are taken.
```
<!-- plain comment -->

<!-- wp:paragraph -->
<p>content of paragraph<br> <!-- comment inside paragraph -->
This is a long sentence just for the sake of testing that there is no change and a line break before the paragraph tag.
</p>
<!-- /wp:paragraph -->

More content without block.
<!-- another comment with <p> tag -->

<script>alert('<p>I am a paragraph!</p>');</script>
<!--nextpage-->
<pre>

<!-- plain comment -->

content of paragraph inside pre<br> <!-- comment inside pre line -->
This is a long sentence just for the sake of testing that there is no change and a line break at the end.

<!-- a comment on a line -->

More content without block.
<!-- another comment -->

</pre>
```

## Types of changes
 Bug fix

### CODE REVIEW

**Comment by mattyrob** [comment]
In my testing I think this still needs work.

Basic content as follows:
```
<!-- wp:paragraph -->
<p>Paragraph <strong>One</strong> starts the content.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Paragraph <em>Two</em> is the middle of the content.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Paragraph <a href="https://www.classicpress.net">Three</a> is the end of the content.</p>
<!-- /wp:paragraph -->
```

If this is loaded in the Visual tab and changed to the Text tab it is changed as follows:
```
&nbsp;

<!-- wp:paragraph -->
Paragraph <strong>One</strong> starts the content.

<!-- /wp:paragraph -->

<!-- wp:paragraph -->


Paragraph <em>Two</em> is the middle of the content.

<!-- /wp:paragraph -->

<!-- wp:paragraph -->


Paragraph <a href="https://www.classicpress.net">Three</a> is the end of the content.

<!-- /wp:paragraph -->
&nbsp;
```

And if this is loaded in the Text tab, checked in Visual and back to the Text tab it produces the following:
```
<!-- wp:paragraph -->


Paragraph <strong>One</strong> starts the content.

<!-- /wp:paragraph -->

<!-- wp:paragraph -->


Paragraph <em>Two</em> is the middle of the content.

<!-- /wp:paragraph -->

<!-- wp:paragraph -->


Paragraph <a href="https://www.classicpress.net">Three</a> is the end of the content.

<!-- /wp:paragraph -->
&nbsp;
```

**Comment by mattyrob** [comment]
Perhaps I'm missing something then but the addition of the line breaks and the `&nbsp;` seems incorrect to me - the contact should not be changing when switching between the two editor screens - should it?

**Comment by joyously** [comment]
There is a separate piece of code that explicitly "keeps empty paragraphs" by adding the `&nbsp;`.
From within the `autop` and `removep` functions, you can't really tell whether the comment is supposed to be part of the paragraph or not. I had tried one version with including the `\n` in the preserved comment, but that version got a JS error when the selection code tries to add a node inside a comment, which you can't do. So I removed that. I still think it's the way to go, though, because the line break of the comment should not influence the algorithm of adding or removing paragraphs.
Any ideas of how to do that better?

**Comment by joyously** [comment]
> the content should not be changing when switching between the two editor screens - should it?

Yes, it should, and it is designed to. `removep` says
```
  * Replaces <p> tags with two line breaks except where the <p> has attributes.
  * Unifies whitespace.
  * Indents <li>, <dt> and <dd> for better readability.
```
and `autop` says
```
Replaces two line breaks with a paragraph tag and one line break with a <br>.
```
The code actually isolates all block-level tags with line breaks, and removes line breaks that are within tags. Also, if a p tag has attributes, it is not changed.

Because the comments might have line breaks, they affect the way the paragraphs are found (expecting `\n\n`), and the comments are surrounded by p tags (which is why WP put that fix in that I removed).
I know it's not perfect, but I need help seeing how I should change it.

**Comment by mattyrob** [comment]
I think we are talking across purposes, those 2 functions are there to ensure consistent display in the Visual and Text tabs, but the two functions cycle and if you being with some content in the Visual tab, switching to the Text view and then back should ensure all content remains the same.

This same process should apply when starting in Text mode and switching to Visual and back.

The current code is not doing that, it is starting with the database content as loaded and it is changing the whitespacing, adding `&nbsp;` entities and stripping out paragraph tags.

I think an issue was raised about the Undo button being activated in the Editor - this happens when the code / browser deems that the content has changed. Our testing goal should be that loading the editor and changing between Visual and Text modes in the editor does not result in activation of the Undo button.

This currently works just fine in the Editor in most situations but specifically it breaks if there is a comment tag at the start of the content and when there are Block Editor comments in the content from the WP editor. Those are the scenarios we need to develop a solution for.

**Comment by joyously** [comment]
Your test data has p tags. 
The block editor saves p tags in the database. The classic editor removes p tags before saving. The PHP `wpautop` function puts the p tags back in `the_content` filter, for the front end.
Switching between modes in the classic editor adds p tags for Visual mode and removes them for Text mode.

It occurs to me that the big problem was the removal of line breaks so the p tags weren't put back correctly, so perhaps there was another subtle change in that last Tiny update that worked for WP but not CP.

**Comment by mattyrob** [comment]
> so perhaps there was another subtle change in that last Tiny update that worked for WP but not CP.

I don;t think so, I can reproduce this same issue in CP 1.3.1 (and as I recall in WP 4.9.20) - so with the version of TinyMCE before the update.

**Comment by mattyrob** [comment]
I've done some testing on this a little more now, I really like the approach of storing and replacing the comment blocks but I think I may have identified a simpler fix.

This line:
`if ( $markerParent.is( 'p' ) && ! $markerParent.children().length && ! $markerParent.text() ) {`

is the check performed to see if a parent tag for a bookmark needs to be stripped, but the last check looks for text and returns null when there is only a comment present. A small change here to:

`if ( $markerParent.is( 'p' ) && ! $markerParent.children().length && ! $markerParent.html() ) {`

This preserved the opening tag, and as far as I can tell preserves all comments in the post content.

That would then leave the issue of Block contant spacing as detailed above for further consideration.

**Comment by joyously** [comment]
In my PR, the `remove` in that if statement is commented out, since it was deleting content that should be kept. But that means it wasn't deleting the empty paragraph that is added just before that check (supposedly for cursor positioning).
Are you saying the `remove` should be put back, with your change to the if statement?

Be aware that `removep` is called just before every `autop`.

**Comment by mattyrob** [comment]
I had noted you'd commented that section which is part of why I checked more deeply what it was doing.

If you take a starting point from current `develop` and make that single change as outline above, all of the other changes seem to become unnecessary in terms of removing and restoring comments.

Currently, in `develop` with no changes, when moving between the two Editor tabs, a 'bookmark' is added to the content to indicate the current cursor location, this is then removed. It is added as a `<span>` tag. The issue we are encountering is that if the very beginning content of the post_content data is an JTML comment field, it is getting stripped out along with this bookmark because TinyMCE does not recognise comments in the same way as text. By looking for `html()` instead of `text()` the comment can be identified and kept in place without any need to remove and then restore all comments.

So, in summary, yes add the `remove()` call back as such but all of the other changes can be replaced by the single change above - the only issue that remains is the spacing of Block data as described above - but the editor 'Undo' and 'Redo' buttons no longer activate with this change.
